src/traces/table/plot.js

```diff
@@ -0,0 +1,892 @@
+    // fixme rowBlock truthiness check is due to ugly hack of placing 2nd panel as d.page = -1
+    var rowFrom = rowBlock ? rowBlock.rows[0].rowIndex : 0;
+    var rowTo = rowBlock ? rowFrom + rowBlock.rows.length : 0;
+    return [rowFrom, rowTo];
+}
+
+function headerHeight(d) {
+    var headerBlocks = d.rowBlocks[0].auxiliaryBlocks;
+    return headerBlocks.reduce(function(p, n) {return p + rowsHeight(n, Infinity);}, 0);
+}
+
+function findPagesAndCacheHeights(blocks, scrollY, scrollHeight) {
+
+    var pages = [];
+    var pTop = 0;
+
+    for(var blockIndex = 0; blockIndex < blocks.length; blockIndex++) {
+
+        var block = blocks[blockIndex];
+        var blockRows = block.rows;
+        var rowsHeight = 0;
+        for(var i = 0; i < blockRows.length; i++) {
+            rowsHeight += blockRows[i].rowHeight;
+        }
+
+        // caching allRowsHeight on the block - it's safe as this function is always called from within the code part
+        // that handles increases to row heights
+        block.allRowsHeight = rowsHeight;
+
+        var pBottom = pTop + rowsHeight;
+        var windowTop = scrollY;
+        var windowBottom = windowTop + scrollHeight;
+        if(windowTop < pBottom && windowBottom > pTop) {
+            pages.push(blockIndex);
+        }
+        pTop += rowsHeight;
+
+        // consider this nice final optimization; put it in `for` condition - caveat, currently the
+        // block.allRowsHeight relies on being invalidated, so enabling this opt may not be safe
+        // if(pages.length > 1) break;
+    }
+
+    return pages;
+}
+
+function updateBlockYPosition(gd, cellsColumnBlock, tableControlView) {
+    // fixme this function is THE performance hotspot
+    var d = cellsColumnBlock[0][0].__data__;
+    var blocks = d.rowBlocks;
+    var calcdata = d.calcdata;
+
+    var bottom = firstRowAnchor(blocks, blocks.length);
+    var scrollHeight = d.calcdata.groupHeight - headerHeight(d);
+    var scrollY = calcdata.scrollY = Math.max(0, Math.min(bottom - scrollHeight, calcdata.scrollY));
+
+    var pages = findPagesAndCacheHeights(blocks, scrollY, scrollHeight);
+    if(pages.length === 1) {
+        if(pages[0] === blocks.length - 1) {
+            pages.unshift(pages[0] - 1);
+        } else {
+            pages.push(pages[0] + 1);
+        }
+    }
+
+    // make phased out page jump by 2 while leaving stationary page intact
+    if(pages[0] % 2) {
+        pages.reverse();
+    }
+
+    cellsColumnBlock
+        .each(function(d, i) {
+            // these values will also be needed when a block is translated again due to growing cell height
+            d.page = pages[i];
+            d.scrollY = scrollY;
+        });
+
+    cellsColumnBlock
+        .attr('transform', function(d) {
+            var yTranslate = firstRowAnchor(d.rowBlocks, d.page) - d.scrollY;
+            return 'translate(0 ' + yTranslate + ')';
+        });
+
+    // conditionally rerendering panel 0 and 1
+    if(gd) {
```

Comment:
How can `gd` be falsy here?

Reply:
`gd` is only needed (and passed on by the caller) if there's a possible need for revolver panel rerender. If the user scrolls down, then maybe one panel jumps ahead of the other, implying that its contents must be rendered. But if this function merely gets called due to a row becoming taller (eg. to fit LaTeX or wrapped text content) then it just bumps the block Y positions but doesn't rerender the block contents.

Should I split this function to two parts?

Reply:
Ok. Not a great pattern, but no action required for now.